samza-kafka/src/main/scala/org/apache/samza/system/kafka/NewKafkaSystemConsumer.java

```diff
@@ -0,0 +1,412 @@
+
+    if (numTPs > 0) {
+      perPartitionFetchThreshold = fetchThreshold / numTPs;
+      LOG.info("perPartitionFetchThreshold=" + perPartitionFetchThreshold);
+      if (fetchThresholdBytesEnabled) {
+        // currently this feature cannot be enabled, because we do not have the size of the messages available.
+        // messages get double buffered, hence divide by 2
+        perPartitionFetchThresholdBytes = (fetchThresholdBytes / 2) / numTPs;
+        LOG.info("perPartitionFetchThresholdBytes is enabled. perPartitionFetchThresholdBytes="
+            + perPartitionFetchThresholdBytes);
+      }
+    }
+  }
+
+  @Override
+  public void stop() {
+    LOG.info("Stopping Samza kafkaConsumer " + this);
+
+    if (!stopped.compareAndSet(false, true)) {
+      LOG.warn("attempting to stop stopped consumer.");
+      return;
+    }
+
+    // stop the proxy (with 5 minutes timeout)
+    if (proxy != null) {
+      LOG.info("Stopping proxy " + proxy);
+      proxy.stop(TimeUnit.MINUTES.toMillis(5));
+    }
+
+    try {
+      synchronized (kafkaConsumer) {
+        LOG.info("Closing kafka consumer " + kafkaConsumer);
+        kafkaConsumer.close();
+      }
+    } catch (Exception e) {
+      LOG.warn("failed to stop SamzaRawKafkaConsumer + " + this, e);
+    }
+  }
+
+  /*
+   record the ssp and the offset. Do not submit it to the consumer yet.
+   */
+  @Override
+  public void register(SystemStreamPartition systemStreamPartition, String offset) {
+    if (started.get()) {
+      String msg =
+          String.format("Trying to register partition after consumer has been started. sn=%s, ssp=%s", systemName,
+              systemStreamPartition);
+      LOG.error(msg);
+      throw new SamzaException(msg);
+    }
+
+    if (!systemStreamPartition.getSystem().equals(systemName)) {
+      LOG.warn("ignoring SSP " + systemStreamPartition + ", because this consumer's system is " + systemName);
+      return;
+    }
+    super.register(systemStreamPartition, offset);
+
+    TopicPartition tp = toTopicPartition(systemStreamPartition);
+
+    topicPartitions2SSP.put(tp, systemStreamPartition);
+
+    LOG.info("Registering ssp = " + systemStreamPartition + " with offset " + offset);
+
+    String existingOffset = topicPartitions2Offset.get(tp);
+    // register the older (of the two) offset in the consumer, to guarantee we do not miss any messages.
+    if (existingOffset == null || compareOffsets(existingOffset, offset) > 0) {
+      topicPartitions2Offset.put(tp, offset);
+    }
+
+    samzaConsumerMetrics.registerTopicAndPartition(toTopicAndPartition(tp));
+  }
+
+  /**
+   * Compare two String offsets.
+   * Note. There is a method in KafkaAdmin that does that, but that would require instantiation of systemadmin for each consumer.
+   * @param off1
+   * @param off2
+   * @return see {@link Long#compareTo(Long)}
+   */
+  public static int compareOffsets(String off1, String off2) {
+    return Long.valueOf(off1).compareTo(Long.valueOf(off2));
```

Comment:
Rather than parsing the offsets here and comparing, it will be better to use the comparator from `kafkaSystemAdmin` here(KafkaSystemAdmin.offsetComparator method).